crates/amaru/src/stages/pull_p2p.rs

```diff
@@ -0,0 +1,348 @@
+            // and keep progressing the chain-sync protocol with this particular peer.
+            //
+            // More sophisticated behaviors might need to decide what to do with the peer depending
+            // on back-pressure from downstream stages that need to process the header.
+            InitiatorEvent::BlockHeaderReceived(pid, header, tip) => {
+                tracing::info!(%pid, "block header received");
+
+                let work = WorkUnit {
+                    sync_data: Some(ChainSyncEvent::RollForward {
+                        peer: pallas_to_amaru_peer(&pid),
+                        raw_header: header.cbor,
+                        point: todo!(),
+                        span: Span::current(),
+                    }),
+                    network_cmd: Some(InitiatorCommand::ContinueSync(pid.clone())),
+                    ..Default::default()
+                };
+
+                WorkSchedule::Unit(work)
+            }
+
+            // This is triggered when a rollback has been received from a peer that is performing
+            // chain-sync. Comments from the block header received event apply here as well.
+            InitiatorEvent::RollbackReceived(pid, point, _) => {
+                tracing::info!(%pid, "rollback received");
+
+                let work = WorkUnit {
+                    sync_data: Some(ChainSyncEvent::Rollback {
+                        peer: pallas_to_amaru_peer(&pid),
+                        rollback_point: pallas_to_amaru_point(&point),
+                        span: Span::current(),
+                    }),
+                    network_cmd: Some(InitiatorCommand::ContinueSync(pid.clone())),
+                    ..Default::default()
+                };
+
+                WorkSchedule::Unit(work)
+            }
+
+            // This is triggered when a block body has been received from a peer that is handling a
+            // block-fetch request. The network machinery explicitly avoids decoding the body for
+            // performance reasons. It's up for us (amaru) to decide when / how to execute the
+            // decoding.
+            //
+            // The downside of this approach is that we don't have much information about
+            // the block at this point, just the payload. We can't infer which point in the chain or
+            // to which downstream request it belongs to. The behavior here is to just dispatch the
+            // block body downstream and let downstream stages decide how to handle it.
+            //
+            // There's no need for explicit commands back to the network machinery since the
+            // continuation of the mini-protocol is unambiguous. If the fetch request involved more
+            // blocks, the network machinery will keep receiving and notifying us of new
+            // block bodies.
+            InitiatorEvent::BlockBodyReceived(pid, body) => {
```

Comment:
What's `pid` here? the `peerId`? This would seem useful to correlate with the request

Reply:
Or maybe not if don't care where the block body came from.